### ISSUE TITLE
Unexpected token '>>' when linting Jupyter notebook with doctest/interpreter snippet

### ISSUE THREAD

**Issue by mattharrison** [astral-sh/ruff#9189]
I'm trying to lint notebooks that have interactive snippets in the cells.

To recreate, simple make a notebook with something like this in the cell

```pycon
>>> def function(a,b):
...    return a+b
>>> function(1,2)
```

Ruff (v 0.1.8) fails with this error:

```
% ruff foo.ipynb
error: Failed to parse foo.ipynb:cell 1:1:1: Unexpected token '>>'
foo.ipynb:cell 1:1:1: E999 SyntaxError: Unexpected token '>>'
Found 1 error.
```

**Comment by dhruvmanila** [comment]
Can you expand on what does "interactive snippets" mean? Is the code snippet you've mentioned to be pasted as it is including the prompt `>>>` and `...`?

**Comment by mattharrison** [comment]
Yes, it means writing code in a doctest style. Jupyter supports this style of code and I would love to be able to format it. Here's a screenshot of my notebook, it is literally a cell with the code I pasted above.


<img width="497" alt="Screen Shot 2023-12-18 at 9 56 19 AM" src="https://github.com/astral-sh/ruff/assets/119405/968837ef-abfa-41f9-9518-1d323c26c280">

**Comment by dhruvmanila** [comment]
TIL, thanks for providing the context. Can you further expand on your use-case for this? I might be misunderstanding you or missing some context but wouldn't using the interactive cells of the notebook itself be better instead? So, what I mean is the following:

<img width="478" alt="Screenshot 2023-12-19 at 10 54 35" src="https://github.com/astral-sh/ruff/assets/67177269/b69b9464-e416-4fe9-970c-bcebb4d81c8b">

**Comment by mattharrison** [comment]
My use case is I write books in Jupyter and want to format the code. As to
why I use interpreter snippets? So the reader can distinguish between code
and output. (And so I can test the code in my book.)

**Comment by dhruvmanila** [comment]
Thanks for providing this information! I think it can be compared to the way Python's official tutorials are written (for example https://docs.python.org/3/tutorial/controlflow.html#if-statements) where the text would be a markdown block while the code snippets would be a code block. 

As for the implementation side, maybe we could utilize some parts of the new docstring formatter to get the code part from the interactive snippets (cc @BurntSushi). We could check if the first line begins with the `>>>` prompt and use an intermediary step to extract the code parts. I can look into it sometime next year (after my Christmas / New Year break) ;)

**Comment by SergejsKims** [comment]
I have similar behavior with Databricks Notebooks. In Databricks it is normal to define `%pip install ...`, but in this case Ruff fails with `Unexpected token '%'`, At least, would be good to Skip such line.

**Comment by dhruvmanila** [comment]
What's the difference between a Databricks Notebook and a Jupyter Notebook? Do they have different file extension? If it's `.ipynb`, then Ruff does support parsing those magic commands (`%pip install ...`).

**Comment by SergejsKims** [comment]
Databricks Notebook has "*.py" extension.

**Comment by dhruvmanila** [comment]
Ruff uses the file extension to determine the expected syntax. If it's a `.py` file, it'll expect a Python source code and similarly, if it's `.ipynb` it expects a JSON encoded Notebook source. I'm not sure why does it uses `.py` extension because any Python parser would fail to parse such files.

**Comment by bthorsted** [comment]
@dhruvmanila Databricks use `.py` because the contents are not JSON encoded, but rather plain text python code with some magic comments to escape features like `%pip install` and to mark cell boundaries.

I assume they made that design choice to make it more git-friendly while still having the benefits of a notebook-style interface.

The magic comments are parsed and rendered by the databricks web interface, so the user won't see that cells are defined by a header-style magic comment: `# COMMAND ----------`.
Lines beginning with `%` in the web editor will be stored as `# MAGIC %` in the raw `.py` file.

**Comment by dhruvmanila** [comment]
> Lines beginning with `%` in the web editor will be stored as `# MAGIC %` in the raw `.py` file.

If this is the case, then it shouldn't create an issue with Ruff as it's just a Python comment. As such, this is different than what the author of this issue is facing.

**Comment by Avasam** [comment]
If I'm not mistaken, I think Ruff now supports `>>>` in docstrings?

Maybe this could be expanded to notebooks?
And I'd like to also have support for stdin (even if it means I have to pass a fake filename+extension through `--stdin-filename`

**Comment by dhruvmanila** [comment]
> If I'm not mistaken, I think Ruff now supports `>>>` in docstrings?

Yeah, that's true. I haven't looked at the code for this but it could be possible to extract it out as a common infra to be used across multiple interfaces.
